### PR TITLE
refactor: cache inverted index with fixed-size page

### PR DESCRIPTION
**Pull request by CookiePieWw**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

#5011

## What's changed and what's your intention?

As title. Currently the cache will pull the whole file.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

File: src/mito2/src/cache/index.rs [hunk 1]
```diff
@@ -59,25 +60,47 @@ where
...
+        for index in indexes {
+            let page = match self.cache.get_index(&index) {
+                Some(page) => {
+                    CACHE_HIT.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
+                    page
+                }
+                None => {
+                    let page = self
+                        .inner
+                        .seek_read(
```
Comment (zhongzc) [inline src/mito2/src/cache/index.rs:81]: The name `read_range` seems more appropriate

File: src/mito2/src/cache/index.rs [hunk 2]
```diff
@@ -59,25 +60,47 @@ where
...
     ) -> index::inverted_index::error::Result<Vec<u8>> {
-        let range = offset as usize..(offset + size as u64) as usize;
-        if let Some(cached) = self.cache.get_index(IndexKey {
-            file_id: self.file_id,
-        }) {
-            CACHE_HIT.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
-            Ok(cached[range].to_vec())
-        } else {
-            let mut all_data = Vec::with_capacity(1024 * 1024);
-            self.inner.read_all(&mut all_data).await?;
```
Comment (zhongzc) [inline src/mito2/src/cache/index.rs:70]: `read_all` can be removed.

Comment (CookiePieWw) [reply src/mito2/src/cache/index.rs:70]: these functions are trait member of `InvertedIndexReader` and used for `InvertedIndexBlobReader` as well, do we need to change them all?

File: src/mito2/src/cache/index.rs [hunk 3]
```diff
@@ -132,6 +148,47 @@ impl<R: InvertedIndexReader> InvertedIndexReader for CachedInvertedIndexBlobRead
...
+
+    /// Ranges of first page.
+    /// For example, if offset is 1000 and size is 5000 and PAGE_SIZE is 4096, then the first page is 1000..4096.
+    fn offset_to_first_range(offset: u64, size: u32, page_size: usize) -> Range<usize> {
+        (offset % (page_size as u64)) as usize
+            ..if size > page_size as u32 - (offset % (page_size as u64)) as u32 {
+                page_size
+            } else {
+                (offset % (page_size as u64) + size as u64) as usize
+            }
```
Comment (zhongzc) [inline src/mito2/src/cache/index.rs:173]: Master of arithmetic

Comment (CookiePieWw) [reply src/mito2/src/cache/index.rs:173]: I'll try to simplify it 😥

File: src/mito2/src/cache/index.rs [hunk 4]
```diff
@@ -59,43 +60,58 @@ where
...
+                    let page = self
+                        .inner
+                        .range_read(
+                            index.page_id * (self.cache.page_size as u64),
+                            self.cache.page_size as u32,
+                        )
+                        .await?;
+                    self.cache.put_index(index.clone(), Arc::new(page.clone()));
+                    CACHE_MISS.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
+                    Arc::new(page)
```
Comment (zhongzc) [inline src/mito2/src/cache/index.rs:88]:
```suggestion
                    let page = Arc::new(page);
                    self.cache.put_index(index.clone(), page.clone());
                    CACHE_MISS.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
                    page
```

Comment (zhongzc) [reply src/mito2/src/cache/index.rs:88]: clone the arc instead of the vec

File: src/mito2/src/cache/index.rs [hunk 5]
```diff
@@ -59,43 +60,58 @@ where
...
+                    &page[IndexKey::offset_to_first_range(offset, size, self.cache.page_size)],
+                );
+            } else if index.page_id == last_page_id {
+                data.extend_from_slice(
+                    &page[IndexKey::offset_to_last_range(offset, size, self.cache.page_size)],
+                );
+            } else {
+                data.extend_from_slice(&page);
+            }
         }
```
Comment (zhongzc) [inline src/mito2/src/cache/index.rs:102]: suggestion: a more concise way
```suggestion
        let last_index = indexes.len() - 1;
        for (i, index) in indexes.into_iter().enumerate() {
            let page = match self.cache.get_index(&index) {
                Some(page) => {
                    CACHE_HIT.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
                    page
                }
                None => {
                    let offset = index.page_id * (self.cache.page_size as u64);
                    let size = self.cache.page_size as _;
                    let page = self.inner.range_read(offset, size).await?;

                    let page = Arc::new(page);
                    self.cache.put_index(index.clone(), page.clone());
                    CACHE_MISS.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
                    page
                }
            };

            let range = if i == 0 {
                // first page range
                IndexKey::offset_to_first_range(offset, size, self.cache.page_size)
            } else if i == last_index {
                // last page range
                IndexKey::offset_to_last_range(offset, size, self.cache.page_size)
            } else {
                0..self.cache.page_size
            };

            data.extend_from_slice(&page[range]);
        }
```

File: src/mito2/src/cache/index.rs (hunk 4 above)
Comment (WenyXu) [inline src/mito2/src/cache/index.rs:88]: We can read all cache-hit pages first, and fetch all cache-missed pages via `RangeReader::read_vec`.

File: src/mito2/src/cache/index.rs [hunk 6]
```diff
@@ -59,43 +60,58 @@ where
...
+                data.extend_from_slice(
+                    &page[IndexKey::offset_to_first_range(offset, size, self.cache.page_size)],
+                );
+            } else if index.page_id == last_page_id {
+                data.extend_from_slice(
+                    &page[IndexKey::offset_to_last_range(offset, size, self.cache.page_size)],
+                );
+            } else {
+                data.extend_from_slice(&page);
+            }
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/mito2/src/cache/index.rs:101]: FYI:  https://github.com/apache/opendal/blob/7cde1f585bc00e42c459a463956b6bc260d2eaaa/core/src/types/buffer.rs#L115-L123

**Review by zhongzc — APPROVED** [review]
great job!

**Review by WenyXu — APPROVED** [review]

**Pull request merged**